### PR TITLE
LT15: add minimum_empty_lines_between_statements

### LINKED ISSUE #6840: [LT15] `minimum_empty_lines_between_statements`
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Thank you for the addition of this new rule! 🙏 

`maximum_empty_lines_between_statements` is very useful, but could we also have `minimum_empty_lines_between_statements`?

### Use case

We want to make sure that there is at least one blank line in between statements, to avoid things like this:

```sql
SELECT ...
FROM ...
WHERE ...;
SELECT ...
FROM ...
WHERE ...;
SELECT ...
FROM ...
WHERE ...;
```

### Dialect

Any

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by adhavan18**
### Brief summary of the change made

Fixes #6840.

`LT15` can cap the blank lines between statements but cannot require any, so there is no way to stop statements being run together:

```sql
SELECT a FROM tab;
SELECT b FROM tab;
SELECT c FROM tab;
```

This adds `minimum_empty_lines_between_statements`, the mirror of the existing `maximum_empty_lines_between_statements`. With it set to `1`:

```
L:   1 | P:  19 | LT15 | Expected at least 1 blank line(s) between statements, found 0.
L:   2 | P:  19 | LT15 | Expected at least 1 blank line(s) between statements, found 0.
```

and `sqlfluff fix` inserts the blank lines.

**It defaults to `0`, so nothing changes for anyone already using this rule.** There is a test pinning that.

### Design notes

A few things were less obvious than the symmetry suggests, so recording the choices:

- **Only checked outside a statement.** That is where "between statements" is meaningful, and it keeps the option away from the `with_compound_statement` branch, which exists to avoid conflicting with LT08's blank-line-after-CTE requirement.
- **Runs before the maximum**, so the two cannot both act on one gap: one wants newlines added, the other wants them removed.
- **Fires on the last newline of a run**, located with `context.siblings_post` (the same approach LT09 uses). Reporting on every newline in a gap would produce one violation per line.
- **Whitespace is stepped over in both directions**, so a line containing only spaces still counts as a blank line. This one came out of a test: the forward lookahead originally skipped only indent/dedent, so a `\n   \n` gap was measured from the wrong newline and reported "found 0" on an already-correct file.
- **Templated newlines and EOF are skipped.** There is nothing to pad before the end of the file, and templated whitespace is not ours to rewrite.

### Are there any other side effects of this change that we should be aware of?

None at the default. The option is opt-in and the existing maximum behaviour is untouched, which the pre-existing LT15 cases cover.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - [x] `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
- [x] Added appropriate documentation for the change.
- [ ] Created GitHub issues for any relevant followup/future enhancements if appropriate.

Six cases added: a failing pair fixed to one blank line, a minimum of two, an already-satisfied file, the default-of-zero no-op, end-of-file, and the whitespace-only line. The rule docstring and `default_config.cfg` are updated too.

```
$ pytest test/rules/
2514 passed, 101 skipped
```

`LT15.py` is at 100% statement coverage, and `ruff check`, `ruff format`, `mypy` and `yamllint` are clean on the changed files.

### AI-assisted contribution disclosure

Per CONTRIBUTING: this PR was written with the help of an AI coding assistant, and I remain responsible for it.

**AI-assisted:** drafting the implementation, the fixtures and this description.

**Verified by me:** confirmed the gap exists and that the default is a genuine no-op by linting with no config present, rather than assuming; checked the fix output byte by byte; found and fixed the whitespace-lookahead bug from a failing fixture rather than adjusting the test; and checked `LT15.py` reaches 100% coverage, since the repo gates on it, folding one unreachable branch into an adjacent condition instead of writing a test for a path that cannot happen.

---
## Summary by cubic
Adds `minimum_empty_lines_between_statements` to LT15 to require blank lines between top‑level statements; previously LT15 could only cap blanks. Defaults to 0 (no change unless configured) and runs before and is capped by the maximum so fixes converge; logs a warning when the cap changes a reported result.

- Scope: Only between statements (including inside batches), never inside statements or between batches. Detects batches structurally (covers T‑SQL and Oracle). Treats `GO` and `/` as non‑statements.
- Same-line statements: Enforced by seeking `statement_terminator` (“SELECT …; SELECT …”). Inserts a newline plus required blanks, preserves trailing comments, and ignores inner semicolons in procedural bodies.
- Templates: Counts only editable newlines. Skips templated newlines and templated terminators. Declines gaps touching template block placeholders and does not rewrite template‑owned whitespace.
- Counting/anchoring: Shares one blank‑line counter for the minimum; the maximum keeps its original windowed eligibility to avoid widening behavior. When a comment splits a gap, anchors only the run ending at the next statement and counts across the intervening comment so existing blanks aren’t hidden. Newlines inside a multi‑line `/* */` comment are the comment’s own text, not blanks, and don’t satisfy the minimum.
- Config/tests: Adds `minimum_empty_lines_between_statements = 0` to default config and rule metadata. Tests pin the default no‑op, same‑line enforcement, batch and template edge cases, all blank‑line‑around‑comment permutations, and convergence when min > max.

<sup>Written for commit 6815500c3154672f044100bcb58ff5ced7f387db. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/sqlfluff/sqlfluff/pull/8282?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/sqlfluff/rules/layout/LT15.py [hunk 1]
```diff
@@ -93,3 +122,62 @@ def _eval(self, context: RuleContext) -> Optional[List[LintResult]]:
...
+            seg
+            for seg in context.siblings_post
+            if not seg.is_type("dedent", "indent", "whitespace")
+        ]
+        if following and following[0].is_type("newline"):
+            return None
+
+        # Skip when nothing but the end of the file follows, so there is no gap to
+        # pad, and when the newline is templated, since that is not ours to rewrite.
+        if context.segment.is_templated or not any(
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/sqlfluff/rules/layout/LT15.py]: Confirmed and fixed in `9bcc382`. You were right:

```
\nSELECT a FROM tab;            ->  1 violation
-- header\nSELECT a FROM tab;   ->  1 violation
```

Both were reported as a gap needing padding with no statement in front of them. The check now requires code in `siblings_pre` as well as `siblings_post`, so a gap needs a statement on both sides. Both cases are quiet, and the real detections are unchanged (two statements still fire, three still fire twice). Fixtures added for the leading blank line and the leading comment.

File: src/sqlfluff/rules/layout/LT15.py [hunk 2]
```diff
@@ -93,3 +122,62 @@ def _eval(self, context: RuleContext) -> Optional[List[LintResult]]:
...
+            seg.is_code for seg in context.siblings_post
+        ):
+            return None
+
+        # Count the run of newlines ending at this one. Two newlines are one blank
+        # line, so the blank count is one less than the run length.
+        run = 1
+        for raw_seg in reversed(context.raw_stack):
+            if raw_seg.is_type("newline"):
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/sqlfluff/rules/layout/LT15.py:242]: This one I could not reproduce as a templating problem, and the reason is worth recording.

I expected `is_templated` to be the discriminator, as it is in the maximum check, so I added the same guard. It changed nothing, because these newlines are not templated. Inspecting them directly:

```
newline @L1 is_templated=False is_literal=True src=slice(20, 21)
newline @L2 is_templated=False is_literal=True src=slice(45, 46)
newline @L3 is_templated=False is_literal=True src=slice(20, 21)   <- same source slice as L1
newline @L4 is_templated=False is_literal=True src=slice(45, 46)   <- same source slice as L2
```

They are genuine source newlines inside the loop body; the loop repeats them, so the same source position appears twice in the rendered file. Neither `is_templated` nor `pos_marker.is_literal()` separates that from an ordinary newline.

I checked what `fix` actually does rather than assume, and it is safe: it inserts inside the loop body, which is a real source position, and the result is still valid Jinja that renders with the blank line present. It is over-eager rather than wrong, since it also pads before the first iteration.

I removed the guard I had added instead of leaving it in: it was unreachable, and it dropped LT15 below the 100% coverage gate, which is how I noticed. Happy to revisit if you can point at a case where a genuinely templated newline reaches this path.

File: test/fixtures/rules/std_rule_cases/LT15.yml [hunk 3]
```diff
@@ -263,3 +263,65 @@ test_fail_extra_blank_in_cte_body:
...
+test_pass_minimum_ignores_end_of_file:
+  # no statement follows, so there is no gap to pad
+  pass_str: |
+    SELECT a FROM tab;
+  configs:
+    rules:
+      layout.newlines:
+        minimum_empty_lines_between_statements: 1
+
+test_pass_minimum_counts_a_whitespace_only_line_as_blank:
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply test/fixtures/rules/std_rule_cases/LT15.yml:321]: Added in `9bcc382`: `test_fail_minimum_pads_a_whitespace_only_gap` takes

```
SELECT a FROM tab;
<three spaces>
SELECT b FROM tab;
```

with `minimum_empty_lines_between_statements: 2` and pins the fixed output, so the whitespace-aware counting is now locked in on the fix path and not only the pass path.

File: src/sqlfluff/rules/layout/LT15.py [hunk 4]
```diff
@@ -93,3 +122,62 @@ def _eval(self, context: RuleContext) -> Optional[List[LintResult]]:
...
             )
         ]
+
+    def _check_minimum(self, context: RuleContext) -> Optional[List[LintResult]]:
+        """Require at least ``minimum_empty_lines_between_statements`` blank lines.
+
+        Fires on the last newline of a run so the whole gap is measured once, rather
+        than once per newline in it.
+        """
+        minimum = self.minimum_empty_lines_between_statements
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/sqlfluff/rules/layout/LT15.py]: Correct, and I am leaving it out of scope deliberately rather than by oversight.

`SELECT a; SELECT b;` on one line has no gap to measure: the check counts blank lines in a run of newlines, and there is no newline at all between those statements. Making it fire would mean inserting a line break where none exists, which is a different operation from padding an existing gap, and it overlaps what a "statements on their own line" rule would do rather than an empty-lines rule.

Given the option is called `minimum_empty_lines_between_statements`, I think reporting "expected at least 1 blank line" on a single-line file would be the wrong message even if the fix were right. Happy to open a separate issue for it if you think the rule should own that case.

**Comment by adhavan18** [comment]
gentle nudge on this one, and a question that is probably the real gate.

all the automated review here is resolved (each finding reproduced, fixed, and pinned by a test), and CI is green. what it hasn't had is a maintainer saying whether the rule/option *should exist*, which is a judgement call rather than something more polish from me can settle.

if the answer is no, i'd genuinely rather know so i can close it. if it's yes but the shape is wrong, happy to change it. and if it's just queue depth, no problem, i'll leave it alone.

for context, my previous sqlfluff PRs (#8163, #8124, #8194, #8240, #8241, #8276) were dialect parsing and test pinning, which are easier calls. this batch is new rules, so it needs an actual decision from someone.

**Comment by adhavan18** [comment]
the P1 is real and reproducible. pushed `6fa2e9d`, which fixes it and two more things it led me to.

**1. same-line statements (the P1).** the crawler only sought `newline`, so there was nothing to crawl between two statements sharing a line and the gap was accepted at any setting:

```
$ sqlfluff lint x.sql   # SELECT a FROM tab; SELECT b FROM tab;
All Finished!           # minimum_empty_lines_between_statements = 1
```

`statement_terminator` is now sought as well. that case needs a different fix from the run-based one, because the line break itself is missing, not just the blank lines, so it creates `minimum + 1` newlines and drops the separating space:

```sql
SELECT a FROM tab;

SELECT b FROM tab;
```

re-linting the fixed file is clean, so it converges in one pass.

**2. templated newlines (the P3).** you were right that this diverged from the max direction. the run count now bails on `is_templated` the way the max branch does, so both directions measure the same editable lines.

**3. a false positive i found while testing that.** this one was mine and neither review caught it. on a file with a jinja block between the statements, the rule fired on the newlines either side of the tag:

```
L:   2 | P:  24 | LT15 | Expected at least 1 blank line(s) between statements, found 0.
L:   3 | P:  13 | LT15 | Expected at least 1 blank line(s) between statements, found 0.
```

and `fix` put blank lines around `{% for %}` / `{% endfor %}`. a block tag sits in the gap as a `placeholder` meta, so that is not a gap between two statements at all. the check now looks for a placeholder in the gap and skips.

**tests**: same-line at minimum 1 and 2, minimum 0 still a no-op on a shared line, terminator at end of file, and the jinja block case. 27 LT15 cases pass; full `test/rules/` suite 2522 passed, 101 skipped.

**Comment by adhavan18** [comment]
the 100% coverage gate failed on my last push and it was worth failing: it caught dead code. pushed `e10f887`.

two lines were uncovered, and the interesting one was the templated-newline guard i added for your P3:

```python
if raw_seg.is_templated:
    return None
```

i couldn't reach it from any test, and the reason is that the placeholder skip i'd added for the jinja block false positive **already handles that case, earlier**. a gap whose newlines come from a template has that template's own placeholder sitting in it, so the skip returns before the run is ever counted. templated lines can't reach the count to satisfy the minimum in the first place.

so the guard was doing nothing except reading as though it were the thing protecting that behaviour. removed, with a comment at the placeholder skip explaining that it covers both.

the concern you raised is still addressed, just at a different point than my first attempt put it, and now there's no second mechanism claiming to do the same job.

the other uncovered line was the templated-terminator skip, which now has a case: a `;` that comes out of a `{% set %}` isn't the user's line to break.

```
88 stmts, 0 miss, 100%
28 LT15 cases pass
test/rules/: 2523 passed, 101 skipped
```

**Comment by adhavan18** [comment]
all four are valid, and the first one is a mistake i should not have made. pushed `95d2812`.

## the templated guard: you're right, and my reasoning was bad

i removed `raw_seg.is_templated` because the 100% coverage gate reported it unreachable and i couldn't build a repro. that was the wrong conclusion from that evidence — unreachable *by my test suite* is not unreachable, and i wrote a code comment asserting the placeholder skip covered the case without ever demonstrating it.

it doesn't. only block tags leave a placeholder:

```
SELECT a FROM tab;{{ "\n\n" }}
SELECT b FROM tab;
```
```
[L:  1, P: 18]   statement_terminator:  ';'
[L:  1, P: 19]   newline:               '\n'      <- templated, no placeholder
[L:  1, P: 19]   newline:               '\n'      <- templated, no placeholder
[L:  3, P:  5]   newline:               '\n'      <- real, anchors the run
```

the anchor is a real newline so `context.segment.is_templated` passes, then the backward walk counts the two templated ones. at `minimum = 3` the guarded version bails and mine flagged and then *fixed*, building the gap partly out of lines that aren't in the file. exactly the divergence from the max branch you flagged in the first place.

restored, and this time there's a case that reaches it.

## trailing comment hides the newline

reproduced, and it's worse than a false positive — it mangles a file that was already correct:

```
$ cat t.sql
SELECT a FROM tab; -- x

SELECT b FROM tab;

$ sqlfluff fix t.sql          # minimum = 1
SELECT a FROM tab;

-- x

SELECT b FROM tab;
```

the scan now skips comments as well as whitespace when looking for the newline that ends the line.

## template tag between same-line statements

fixed as you suggested: a `placeholder` before any newline means this isn't two statements sharing a line, so the same-line path returns.

## the terminator branch guard

added. one correction to the finding, since it changes what the test can prove: **mysql doesn't nest those terminators.** `CREATE PROCEDURE p() BEGIN SELECT 1; SELECT 2; END;` parses flat, with both terminators as file-level children, so `parent_stack` has no statement and the guard is a no-op there. i checked four dialects:

| dialect | terminators nested inside a statement |
|---|---|
| tsql | 3 |
| mysql | 0 |
| bigquery | 0 |
| exasol | 0 |

so the case is real but reachable on tsql, and that's what the test uses.

worth flagging separately: because mysql parses flat, `CREATE PROCEDURE ... BEGIN SELECT 1; SELECT 2; END;` there *does* still get linted as top-level gaps, and the fix would split the procedure body. that's not something the guard you suggested can fix, and i'd rather raise it than quietly leave it. happy to handle it here or in a follow-up, whichever you prefer.

```
94 stmts, 0 miss, 100%
32 LT15 cases pass
test/rules/: 2527 passed, 101 skipped
ruff check + format clean
```

**Review by rubytobi — CHANGES_REQUESTED** [review]
Thanks for tackling this and working through the implementation.

Looking at the existing logic for the `maximum _*` rule, it distinguishes three scopes. Meaning is it inside a statement, between statements within a batch, or between batches at file level. The last only for dialects with a batch grammar node, currently that's T-SQL and Oracle only.

The new `minimum_*` rules doesn't factor this in yet. For consistency, we should follow the same approach here.

**Comment by adhavan18** [comment]
@rubytobi thanks, that's the right call and it was a real gap. pushed `76d2b7c`.

## the scope resolution

the minimum only distinguished *inside a statement* from *everything else*, so in a batch dialect it fired on the file-level gap too. on `tsql`:

```
L:   1 | P:  10 | LT15 | Expected at least 1 blank line(s) between statements   <- inside batch, correct
L:   2 | P:  10 | LT15 | Expected at least 1 blank line(s) between statements   <- inside batch, correct
L:   3 | P:   3 | LT15 | Expected at least 1 blank line(s) between statements   <- between batches, wrong
```

that third one is `maximum_empty_lines_between_batches`' scope, and `minimum_empty_lines_between_statements` has no business there. now resolved through one helper used by both the newline and terminator paths, mirroring the maximum's three tiers.

one deliberate difference from the maximum: the file-level test is **structural** (does this file contain `batch` nodes) rather than `dialect.name == "tsql"`. oracle parses `file > batch > statement` exactly like tsql, so the name check would have missed it, and a dialect that gains batch grammar later is handled without a code change.

that also means the maximum's `elif context.dialect.name == "tsql"` currently doesn't apply `maximum_empty_lines_between_batches` on oracle. i haven't touched it — separate behaviour change, and it's yours to call — but flagging it since it's the same blind spot.

## a second thing it turned up

chasing your point exposed a bug i'd have shipped otherwise. the gap checks used `is_code`, and a batch delimiter is code:

| dialect | delimiter | parses as | is a statement? |
|---|---|---|---|
| tsql | `GO` | `go_statement` | no |
| oracle | `/` | `slash_buffer_executor` | no |

so the rule read the delimiter as the next statement and demanded a blank line in front of it — it would have reformatted every `GO` in a file. both sides of the gap are now tested on the `statement` type instead, which is what "between statements" should have meant from the start.

## tests

four new cases, both dialects, both directions: batches adjacent must pass, statements inside a batch must still fail and fix.

```
100 stmts, 0 miss, 100%
36 LT15 cases pass
test/rules/: 2531 passed, 101 skipped
ruff check + format clean
```

the earlier fixes still hold — same-line still caught, trailing comment, jinja block and template-tag cases all still clean.

**Comment by adhavan18** [comment]
﻿found a defect in this PR while re-reading it, and it needs a decision from you rather than just a patch, so flagging the choice as well as the fix. pushed `9886c89`.

## `fix` does not converge when the minimum exceeds the maximum

both settings validate as `range(1000)` independently, so `minimum_empty_lines_between_statements = 5` with `maximum_empty_lines_between_statements = 2` is accepted config. enforced as written the two fixers undo each other, and running `fix` repeatedly alternates forever:

```
pass 1: blank lines = 2  | Expected at least 5 blank line(s) between statements
pass 2: blank lines = 5  | Too many consecutive blank lines
pass 3: blank lines = 2  | Expected at least 5 blank line(s) between statements
pass 4: blank lines = 5  | Too many consecutive blank lines
```

the file never settles, so its contents depend on how many times the fixer happened to run. the in-pass loop protection catches it within a single run (`One fix for LT15 not applied, it would re-cause the same error`), which is why it does not hang, but it does not help across runs and `fix` still exits 0 reporting `FIXED` on a file that is still failing.

this is mine: the minimum is what this PR adds, so the contradiction did not exist before it.

## what i did, and the part that is your call

capped the minimum at the maximum, so `fix` converges immediately and the maximum, being the stricter and longer-standing of the two, keeps control of the boundary they disagree about.

i picked that over the alternative deliberately but i am not confident it is what you want. the options as i see them:

1. **cap** (what i pushed) - convergent, no crash, but silently enforces something other than what was configured
2. **raise `SQLFluffUserError`** - semantically the honest answer, but there is no cross-field validation hook, so raising from `_eval` surfaces as `CRITICAL ... threw an Exception` with a traceback, which reads like a rule crash rather than a config mistake
3. **add cross-field validation to the framework** - `_validate_configs` is per-option today, so this is a core change well outside this PR

happy to switch to 2 or open a separate PR for 3 if either is what you would rather have. leaving it at 1 for now so the branch is at least not shipping a non-convergent fixer.

## tests

two cases added: the min-above-max config settling at the maximum, and the boundary case where min equals max is still satisfiable. the first fails on the parent commit and passes here.

full `test/rules/` suite: **2533 passed, 101 skipped**. `ruff check` and `ruff format --check` clean.

**Comment by rubytobi** [comment]
Functionality wise we are mostly there. One design question I have left: is the below supported?
```sql
-- pre fix
SELECT a FROM tab; SELECT b FROM tab;

-- post fix - current not possible
SELECT a FROM tab;
SELECT b FROM tab;

-- post fix - current possible
SELECT a FROM tab;

SELECT b FROM tab;
```

----

Let's clean up the implementation logic for maintainability going forward, the code tripled in size:

* You made the right call here to have the old `maximum_*` config option take precedence over the new one. Please add a warning for the user when minimum is larger than maximum _and_ this actually affects the lint results. 

* Extend the rule docstring to cover the min/max-capping behavior.

* From a user perspective I'd expect `maximum_*` and `minimum_*` to count the same. There are currently two different ways of counting.

* Replace the duplicate scope logic in `_eval`'s maximum-selection block and `_in_between_statements_scope()` with one shared method both paths call. This also fixes the pre-existing Oracle batch-boundary bug (we currently check for `context.dialect.name == "tsql"` only) on the `maximum` side.

* Consolidate `_check_minimum` / `_check_minimum_same_line` into a shared helper with a unified logic.

**Comment by adhavan18** [comment]
﻿all five implementation points done in `f3454bd`, and the design question answered below. full `test/rules/` suite: **2535 passed, 101 skipped**.

## the design question

> is `SELECT a; SELECT b;` -> a plain line break (no blank line) supported?

no, and i don't think this setting should be what enables it. `minimum_empty_lines_between_statements = 0` is the default, so making 0 mean "split shared lines" would change behaviour for every existing user who never set it. the name is also about blank lines rather than line breaks, so it would be doing something its name doesn't say.

that output is a separate concern - "each statement starts its own line" - and it applies whether or not blank lines are wanted. happy to add it as its own option here (`statements_on_separate_lines`, say) in a follow-up, or leave it out entirely if you'd rather not grow this rule further. i'd rather not fold it into the minimum.

## the maintainability points

**scope logic is now one method.** `_resolve_scope()` returns one of three `_Scope` members and both directions call it: the maximum to pick its limit, the minimum to decide whether it applies at all.

that fixes the oracle bug you pointed at. the file-level test was `dialect.name == "tsql"`, so oracle's batch boundaries were measured against the between-statements maximum:

```
                    # oracle, 3 blank lines at a batch boundary,
                    # maximum_empty_lines_between_batches = 1
before this commit: All Finished!
after:              L:4 | LT15 | Too many consecutive blank lines.
                    L:5 | LT15 | Too many consecutive blank lines.
```

testing for the batch node rather than the dialect name also means a dialect that gains batch grammar later needs no change here. two cases added: the oracle boundary, and a control showing an in-batch gap still uses the between-statements allowance.

**counting is one method.** `_count_blank_lines()` serves both, so they cannot disagree about what a blank line is.

one wrinkle worth your opinion. unifying it initially broke `test_fail_macros_with_newlines`: my run walk aborted on a templated newline anywhere in the run, where the old maximum only inspected a fixed `maximum + 1` window and so never saw the templated one further back. rather than keep two walks, the counter now returns the count **and** whether the run was cut short by templated output. the maximum ignores that flag, since it only ever deletes a newline that is genuinely in the source. the minimum honours it and declines the gap, because padding one whose rendered form already contains lines the source does not would be guessing at the result.

that is one counting implementation with one asymmetry in what the two directions do with it, and the asymmetry is real rather than accidental. if you'd rather the minimum also just counted up to the templated newline and padded, that's a one-line change and i'll make it.

**the two minimum checks are one method.** `_check_minimum(context, *, shares_line=...)`. the shared parts - the gap test, the statement-on-both-sides test, the placeholder skips - live in `_is_gap_between_statements()`; the only real difference left is that a shared line has no run to measure and needs the line break created as well as the blank lines.

**the cap now warns and is documented.** it fires only where it changes a reported result, so a contradictory config that never meets two statements stays quiet:

```
WARNING [LT15] minimum_empty_lines_between_statements (5) is greater than
maximum_empty_lines_between_statements (2); enforcing 2. ...
```

and there's a `.. note::` in the rule docstring covering the capping behaviour and why it exists.

the previous red checks were mine and are fixed: `pre-commit` was a trailing newline in the fixture file.

**Comment by adhavan18** [comment]
﻿both P2s reproduced and fixed in `e07c715`, and the first one is mine: unifying the counter changed the maximum's behaviour, which is exactly what you were asking about when you said the two should count the same way.

## the maximum lost its own eligibility test

@cubic-dev-ai is right. the original check is not just a count:

```python
for raw_seg in context.raw_stack[-maximum_empty_lines - 1 :]:
    if raw_seg.is_templated or not raw_seg.is_type("newline"):
        return None
```

it suppresses the violation outright when the window is not purely newlines. `_count_blank_lines` instead skips whitespace and counts the editable newlines ahead of a templated one, so sharing it widened the rule. measured on a blank run containing one whitespace-only line, `maximum_empty_lines_between_statements = 1`:

```
main:            1 violation
shared counter:  2 violations
after this fix:  1 violation
```

so i have put the maximum back on its own check and left the shared helper with the minimum. that is a partial walk-back of the "count the same way" point, and i would rather say so than quietly not do it: the two directions genuinely do not want the same eligibility rule. the minimum has to treat a line of spaces as blank, because it is deciding whether to add a line the user can see. the maximum declines to touch a run it cannot cleanly own, which is the safer default for a rule that deletes. what they now share is the definition of "how many blank lines is this gap", which is the part that was actually inconsistent.

if you would rather have full unification and are happy to accept the widened maximum as a deliberate change, that is a small edit and i will make it, but it wants to be a decision rather than a side effect. a regression case pins the current behaviour either way.

## a comment split one gap into two fixes

also real:

```
SELECT a FROM tab;
-- a note
SELECT b FROM tab;
```

a comment breaks the gap into two runs of newlines and both qualified, since each has a statement behind it and one ahead. with `minimum_empty_lines_between_statements = 1` that reported twice for one gap, against once for the same file without the comment. only the run that reaches the next statement without passing a comment is the anchor now.

## the two P3s

both fair. `test_pass_minimum_above_maximum_is_capped_not_oscillating` and `test_pass_minimum_equal_to_maximum_is_satisfiable` supplied `fail_str`/`fix_str`, so they are `test_fail_*` now, and the last case has its separating blank line. worth noting i reintroduced the same naming inversion in the regression test i wrote for it, and caught it on the way back through.

## testing

```
test/rules/   2538 passed, 101 skipped
LT15          43 passed, coverage 100%
```

`ruff check` and `ruff format --check` clean.

**Comment by adhavan18** [comment]
﻿@cubic-dev-ai right again, and this one was mine: anchoring past the comment was correct, but the counter still stopped at it, so a blank line on the far side of the comment was invisible and the gap got padded a second time. fixed in `e1a8bf0`.

reproduced before changing anything, `minimum_empty_lines_between_statements = 1`:

```
input                          before this fix                after
---------------------------    ---------------------------    ---------------------------
SELECT a FROM tab;             SELECT a FROM tab;             SELECT a FROM tab;
                                                                                          
-- a note                      -- a note                      -- a note
SELECT b FROM tab;                                            SELECT b FROM tab;
                               SELECT b FROM tab;
                               1 violation, blank added       0 violations, file unchanged
```

the walk now banks the run that ends at a comment and keeps going, because the blank lines on the far side belong to the same gap. a comment still occupies a line, so it ends a run of blank lines rather than counting as one, which keeps `SELECT a; / -- note / SELECT b;` at zero blank lines and correctly gaining one.

all three permutations checked:

| gap | violations | result |
| --- | --- | --- |
| blank line before the comment | 0 | unchanged, already satisfied |
| no blank line either side | 1 | gains exactly one, after the comment |
| blank line after the comment | 0 | unchanged, already satisfied |

## testing

```
test/rules/   2539 passed, 101 skipped
LT15          44 passed, coverage 100%
```

`ruff check` and `ruff format --check` clean. the `rs-build-macos` failure is unrelated to this diff: the runner could not resolve `api.github.com` (`getaddrinfo ENOTFOUND`) while building the rust wheels, and this PR touches only python.

**Comment by rubytobi** [comment]
@adhavan18 is the latest cubic review something you can reproduce as a test case?

**Comment by adhavan18** [comment]
Yes — reproduced it and fixed it in `17f8228c`.

The root cause: a multi-line `/* */` comment lexes as one `block_comment` segment per physical line, with its own newlines between them. `_count_blank_lines` walked those the same as any newline, so a blank line *inside* a comment's own text satisfied the minimum even with zero real blank lines between the statements:

```sql
SELECT a FROM tab;
/* line1

line2 */
SELECT b FROM tab;
```

Confirmed empirically before touching anything: this returns 0 violations on the unfixed rule with `minimum_empty_lines_between_statements: 1`, despite the two statements having no visible separation outside the comment.

Fix distinguishes a `block_comment` segment that *continues* an already-open comment (its raw text doesn't start with `/*`) from one that opens a new one, and skips newlines while inside an unclosed comment body instead of counting them. Scoped to `block_comment` only — `inline_comment` (`--`) never spans multiple segments, and applying the same check to it broke the existing "blank line between two separate `--` comments" case in testing, so that's explicitly excluded.

Added a fail case reproducing this exactly, plus two pass cases guarding the behavior you and cubic already established (blank line between two separate comments, blank line before a multi-line comment) — both still correctly pass. Full LT15 suite (47 cases) and the broader layout family (957 cases) green. Confirmed the new fail case genuinely fails against the unfixed rule before I called it done.

**Comment by adhavan18** [comment]
Good catch — added in `6815500c`: `test_pass_minimum_counts_a_blank_line_after_an_intervening_comment`, the third permutation alongside the existing zero-blank and blank-before cases. All three now have their own fixture.

**Review by rubytobi — APPROVED** [review]
Thanks @adhavan18. 

Just a general observation, we are welcoming any contributor to use AI and help craft PRs. However, we do expect you to be across the changes and understand what you're putting forward. 

Given the numerous back and forth with me and cubic, I'm not confident this was necessarily the case here. Please keep that in mind for future PRs.
